Wrappers/Python/cil/optimisation/algorithms/FISTA.py

```diff
@@ -1,127 +1,273 @@
+    g : Function
+        Convex function with *simple* proximal operator
+    step_size : positive :obj:`float`, default = None
+                Step size for the gradient step of ISTA.
+                The default :code:`step_size` is :math:`\frac{0.99 * 2}{L}.`
+    kwargs: Keyword arguments
+        Arguments from the base class :class:`.Algorithm`.
+
+
+    Examples
+    --------
+
+    .. math:: \underset{x}{\mathrm{argmin}}\|A x - b\|^{2}_{2}
+
+    >>> f = LeastSquares(A, b=b, c=0.5)
+    >>> g = ZeroFunction()
+    >>> ig = Aop.domain
+    >>> ista = ISTA(initial = ig.allocate(), f = f, g = g, max_iteration=10)
+    >>> ista.run()
+
+
+    See also
+    --------
+
+    :class:`.FISTA`
+    :class:`.GD`
+
+
+    """
+
+    @property
+    def provable_convergence_condition(self):
+        return self.step_size <= 0.99*2.0/self.f.L
+
+    @property
+    def step_size(self):        
+       return self._step_size
+
+    # Set default step size
+    def set_step_size(self, step_size):
+        """ Set default step size.
+        """
+        if step_size is None:
+            if isinstance(self.f.L, Number):
+                self._step_size = 0.99*2.0/self.f.L
             else:
-                raise ValueError('{} received both initial and the deprecated x_init parameter. It is not clear which one we should use.'\
-                    .format(self.__class__.__name__))
+                raise ValueError("Function f is not differentiable")
+        else:
+            self._step_size = step_size            
         
-        if initial is not None and f is not None:
-            self.set_up(initial=initial, f=f, g=g)
+    def __init__(self, initial, f, g, step_size = None, **kwargs):
 
-    def set_up(self, initial, f, g=ZeroFunction()):
-        '''initialisation of the algorithm
+        super(ISTA, self).__init__(**kwargs)
+        self._step_size = None
+        self.set_up(initial=initial, f=f, g=g, step_size=step_size, **kwargs)
 
-        :param initial: Initial guess ( Default initial = 0)
-        :param f: Differentiable function
-        :param g: Convex function with " simple " proximal operator'''
+    def set_up(self, initial, f, g, step_size, **kwargs):
+        """ Set up of the algorithm
+        """
 
-        logging.info("{} setting up".format(self.__class__.__name__, ))
-        
-        self.y = initial.copy()
+        # set up ISTA      
+        self.initial = initial
         self.x_old = initial.copy()
-        self.x = initial.copy()
-        self.u = initial.copy()
-
+        self.x = initial.copy()           
         self.f = f
         self.g = g
-        if f.L is None:
-            raise ValueError('Error: Fidelity Function\'s Lipschitz constant is set to None')
-        self.invL = 1/f.L
-        self.t = 1
-        self.configured = True
+
+        # set step_size
+        self.set_step_size(step_size=step_size)
+
+        logging.info("{} setting up".format(self.__class__.__name__, ))
```

Comment:
This should be the 1st line in the method.

Reply:
Fixed